### ISSUE TITLE
False positive: Call to inArray() with different object instances will always evaluate to true

### ISSUE THREAD

**Issue by simPod** [phpstan/phpstan-webmozart-assert#142]
```php
class A {
	public function __construct(private string $val) {
	}	
}

function foo(A $a) {
    Assert::inArray($a, [new A('a')]);
}
```

gives:

> Call to static method Webmozart\Assert\Assert::inArray() with arguments A and array(A) will always evaluate to true

**Comment by ondrejmirtes** [comment]
Yeah, this is unfortunate, it means that this code shouldn't exist https://github.com/phpstan/phpstan-src/blob/3c3ea2f21898e8e411b0a7263910722738839288/src/Rules/Comparison/ImpossibleCheckTypeHelper.php#L85-L134 and all of that should be handled directly in https://github.com/phpstan/phpstan-src/blob/master/src/Type/Php/InArrayFunctionTypeSpecifyingExtension.php.

**Comment by staabm** [comment]
tried working on it, but I don't get how to formulate the unit test to reproduce the problem (I don't get how the `LegacyNodeScopeResolverTest` works for this case)

**Comment by ondrejmirtes** [comment]
You'd need to:

1) Create a new test case that looks like https://github.com/phpstan/phpstan-src/blob/master/tests/PHPStan/Rules/Comparison/ImpossibleCheckTypeStaticMethodCallRuleTest.php and tests the same rule
2) Create a new type-specifying extension that mimics what `in_array` does. You can copy and reduce this file to a minimal working example: https://github.com/phpstan/phpstan-webmozart-assert/blob/978e524796e5e81851d09d5967b173cda173d422/src/Type/WebMozartAssert/AssertTypeSpecifyingExtension.php#L404-L413
3) Create `foo.neon` that registers this extension
4) Return the path to `foo.neon` in the testcase's (from 1)) method `public static function getAdditionalConfigFiles()`.

**Comment by p4veI** [comment]
Hello, I just came across a similar issue as per the following example:

```
/**
 * @param mixed $value
 * @param array<mixed> $array 
 */
public function test($value, array $array) : void
{
  Assert::inArray($value, $array);
}
```

phpstan reports:
`Call to static method Webmozart\Assert\Assert::inArray() with mixed and array will always evaluate to true.`

However this is not true, the webmozart assert function is using the exact same param annotations.

I'm not completely sure if this is related to what @simPod has reported.

**Comment by ondrejmirtes** [comment]
@p4veI In fact this is exactly the same issue :) In fact this issue was moved from phpstan-webmozart-repo here because it needs to be fixed in PHPStan's core.

**Comment by p4veI** [comment]
imo fixed in latest release ;)

**Comment by staabm** [comment]
Please send a regression test, or it didn't happen :-)

**Comment by herndlm** [comment]
Interesting, I just tried to add a regression test for this. The current state is
- the [original snippet](https://github.com/phpstan/phpstan-webmozart-assert/issues/142#issue-1276742494) is *still* reported and therefore not fixed
- the [snippet from @p4veI ](https://github.com/phpstan/phpstan-webmozart-assert/issues/142#issuecomment-1160296916) is not reported and therefore fixed, I assume

**Comment by ondrejmirtes** [comment]
Not surprising - there's still a massive codeblock about `in_array` in ImpossibleCheckTypeHelper: https://github.com/phpstan/phpstan-src/blob/a5f8d2dd0095a80f69fc23364f9cddae67ed7c44/src/Rules/Comparison/ImpossibleCheckTypeHelper.php#L80-L138

It's there to prevent false-positives about `in_array` calls, but it's not going to work if an extension delegates its method call to `in_array` extension via TypeSpecifier.

My comment still stands https://github.com/phpstan/phpstan-webmozart-assert/issues/142#issuecomment-1160296906

**Comment by herndlm** [comment]
ah right, I was just super confused because I could *not* reproduce it in a phpstan snippet. actually it never reported even when it definitely should. that explains it all :) thx
Maybe I try to look at the `in_array` stuff in phpstan soon. that feels more like my domain :)

**Comment by ondrejmirtes** [comment]
Enjoy :D https://github.com/phpstan/phpstan/issues/6705

**Comment by VincentLanglet** [comment]
Just encounter the issue with
```
Assert::inArray('from__r', $fields);
Assert::inArray('to__r', $fields);
Assert::inArray('fromEnd__r', $fields);
Assert::inArray('toEnd__r', $fields);
```
Only the second, third and fourth check are reported as always true.

**Comment by melkamar** [comment]
I have just encountered something similar that I think may have the same root cause?

```php
use Webmozart\Assert\Assert;

class C
{
    /**
     * @param array<string> $supportedNames
     * @param array<string> $enabledNames
     */
    public function __construct(
        array $supportedNames,
        array $enabledNames,
    )
    {
        Assert::allInArray($enabledNames, $supportedNames);
    }
}
```

```
 ------ --------------------------------------------------------------------------------------------------------------------------------
  Line   test.php
 ------ --------------------------------------------------------------------------------------------------------------------------------
  17     Call to static method Webmozart\Assert\Assert::allInArray() with array<string> and array<string> will always evaluate to true.
         🪪  staticMethod.alreadyNarrowedType
 ------ --------------------------------------------------------------------------------------------------------------------------------
```

... commenting just so that this error message ends up in google as it took me a bit of time to find this.

**Comment by ondrejmirtes** [comment]
@melkamar Please run `vendor/bin/phpstan diagnose` and post the output.

I believe this is long fixed and just needs a regression test.

**Comment by melkamar** [comment]
Alright, output below along with the minimal config that reproduces this. I tried using the playground but I don't think the extensions are supported there?

```
$ vendor/bin/phpstan diagnose
Note: Using configuration file /Users/melka/work/rh/phpstan.neon.
PHP runtime version: 8.3.22
PHP version for analysis: 8.3.22 (from runtime)

PHPStan version: 2.1.29
PHPStan running from:
/Users/melka/work/rh/vendor/phpstan/phpstan

Extension installer: Not installed

Included configs from Composer packages:
phpstan/phpstan-webmozart-assert (extension.neon): 2.0.0

Discovered Composer project root:
/Users/melka/work/rh

$ vendor/bin/phpstan analyse -c phpstan.neon test.php
 1/1 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ --------------------------------------------------------------------------------------------------------------------------------
  Line   test.php
 ------ --------------------------------------------------------------------------------------------------------------------------------
  17     Call to static method Webmozart\Assert\Assert::allInArray() with array<string> and array<string> will always evaluate to true.
         🪪  staticMethod.alreadyNarrowedType
 ------ --------------------------------------------------------------------------------------------------------------------------------



 [ERROR] Found 1 error

$ cat phpstan.neon
includes:
    - vendor/phpstan/phpstan-webmozart-assert/extension.neon
parameters:
    level: 4  # lowest level where this starts happening
```

**Comment by ondrejmirtes** [comment]
@melkamar Alright, that's a bug, but probably a different one than in OP. Please create a new issue.
